packages/lime-mesh-upgrade/files/usr/lib/lua/lime-mesh-upgrade.lua

```diff
@@ -0,0 +1,234 @@
+        -- trigger firmware download from master_node url
+        uci:set('mesh-upgrade', 'main', 'upgrade_state', newstate)
+        uci:save('mesh-upgrade')
+        uci:commit('mesh-upgrade')
+        return true
+    end
+    --todo: verify other states 
+    -- lets allow all types of state changes. 
+    uci:set('mesh-upgrade', 'main', 'upgrade_state', newstate)
+    uci:save('mesh-upgrade')
+    uci:commit('mesh-upgrade')
+    return false
+end
+
+-- set download information for the new firmware from master node
+-- Called by a shared state hook in non master nodes
+function mesh_upgrade.set_mesh_upgrade_info(upgrade_data, upgrade_state, transaction_state)
+    local uci = config.get_uci_cursor()
+    if (type(upgrade_data.id) == "number") and string.match(upgrade_data.data.repo_url, "https?://[%w-_%.%?%.:/%+=&]+") ~=
+        nil -- perform aditional checks
+    then
+        if (mesh_upgrade.change_state(upgrade_state)) then
+            uci:get('mesh-upgrade', 'main', 'repo_url')
+            uci:set('mesh-upgrade', 'main', "mesh-upgrade")
+            uci:set('mesh-upgrade', 'main', 'id', upgrade_data.id)
+            uci:set('mesh-upgrade', 'main', 'repo_url', upgrade_data.data.repo_url)
+            uci:set('mesh-upgrade', 'main', 'firmware_ver', upgrade_data.data.firmware_ver)
+            --uci:set('mesh-upgrade', 'main', 'upgrade_state', upgrade_state) already done in change state
+            uci:set('mesh-upgrade', 'main', 'error', 0)
+            uci:set('mesh-upgrade', 'main', 'timestamp', os.time())
+            uci:set('mesh-upgrade', 'main', 'master_node', upgrade_data.master_node)
+            uci:set('mesh-upgrade', 'main', 'transaction_state', transaction_state or mesh_upgrade.transaction_states
+                .STARTED)
+            uci:save('mesh-upgrade')
+            uci:commit('mesh-upgrade')
+            -- trigger shared state data refresh
+            mesh_upgrade.trigger_sheredstate_publish()
+            mesh_upgrade.start_node_download(upgrade_data.data.repo_url)
+        else
+            config.log("invalid state change ")
+        end
+    else
+        config.log("upgrade failed due input data errors")
+    end
+end
+
+--
+-- {
+--     type= "upgrade",
+--     data={
+--       firmware_ver="xxxx",
+--       repo_url="http://10.13.0.1/lros/api/v1/",
+--       upgrde_state="starting,downloading|ready_for_upgrade|upgrade_scheluded|confirmation_pending|~~confirmed~~|updated|error",
+--       error="CODE",
+--       safe_upgrade_status="",
+--       eup_STATUS="",eup.STATUS_DEFAULT = 'not-initiated' eup.STATUS_DOWNLOADING = 'downloading' eup.STATUS_DOWNLOADED = 'downloaded' eup.STATUS_DOWNLOAD_FAILED = 'download-failed'
+--     },
+--     timestamp=231354654,
+--     id="",
+--     transaction_state="started/aborted/finished",
+--     master_node=""
+-- }
+--
```

Comment:
commented code

Reply:
this is just a remainder of the main data type, i found it useful. I think that due to the complexity of this module I can be a good idea to add a readme like other modules have.